File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticConstraint.java
package com.foundgine.core.semantic;

public record SemanticConstraint(SemanticConstraintKind kind, String value, java.math.BigDecimal minimum, java.math.BigDecimal maximum) {
    public static SemanticConstraint range(java.math.BigDecimal min, java.math.BigDecimal max) { return new SemanticConstraint(SemanticConstraintKind.RANGE, null, min, max); }
    public static SemanticConstraint pattern(String pattern) { return new SemanticConstraint(SemanticConstraintKind.PATTERN, pattern, null, null); }
    public static SemanticConstraint temporal(String semantics) { return new SemanticConstraint(SemanticConstraintKind.TEMPORAL, semantics, null, null); }
    public static SemanticConstraint currency(String code) { return new SemanticConstraint(SemanticConstraintKind.CURRENCY, code, null, null); }
    public static SemanticConstraint countryCode(String code) { return new SemanticConstraint(SemanticConstraintKind.COUNTRY_CODE, code, null, null); }
}

File: src/csharp/Foundgine.Core/Semantic/SemanticConstraint.cs
namespace Foundgine.Core.Semantic;

/// <summary>Provider-neutral semantic constraint attached to a field.</summary>
public sealed record SemanticConstraint(
    SemanticConstraintKind Kind,
    string? Value = null,
    decimal? Minimum = null,
    decimal? Maximum = null)
{
    public static SemanticConstraint Range(decimal? minimum = null, decimal? maximum = null) =>
        new(SemanticConstraintKind.Range, Minimum: minimum, Maximum: maximum);

    public static SemanticConstraint Pattern(string pattern) =>
        new(SemanticConstraintKind.Pattern, pattern);

    public static SemanticConstraint Temporal(string semantics) =>
        new(SemanticConstraintKind.Temporal, semantics);

    public static SemanticConstraint Currency(string currencyCode) =>
        new(SemanticConstraintKind.Currency, currencyCode);

    public static SemanticConstraint CountryCode(string countryCode) =>
        new(SemanticConstraintKind.CountryCode, countryCode);
}

public enum SemanticConstraintKind : byte
{
    Range,
    Pattern,
    Temporal,
    Currency,
    CountryCode
}
